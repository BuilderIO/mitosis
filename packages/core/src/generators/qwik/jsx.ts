import { MitosisNode } from '../../types/mitosis-node';
import { DIRECTIVES } from './directives';
import {
  File,
  INDENT,
  invoke,
  NL,
  SrcBuilder,
  quote,
  UNINDENT,
  lastProperty,
} from './src-generator';
import { CssStyles } from './styles';

export function renderJSXNodes(
  file: File,
  directives: Map<string, string>,
  handlers: Map<string, string>,
  children: MitosisNode[],
  styles: Map<string, CssStyles>,
  parentSymbolBindings: Record<string, string>,
  root = true,
): any {
  return function(this: SrcBuilder) {
    if (children.length == 0) return;
    if (root) this.emit('(', INDENT, NL);
    const needsFragment = root && children.length > 1;
    file.import(file.qwikModule, 'h');
    if (needsFragment) {
      file.import(file.qwikModule, 'Fragment');
      this.jsxBeginFragment(file.import(file.qwikModule, 'Fragment'));
    }
    children.forEach((child) => {
      if (isEmptyTextNode(child)) return;
      if (isTextNode(child)) {
        if (child.bindings._text !== undefined) {
          this.jsxTextBinding(child.bindings._text);
        } else {
          this.isJSX
            ? this.emit(child.properties._text)
            : this.jsxTextBinding(quote(child.properties._text!));
        }
      } else {
        let childName = child.name;
        const directive = DIRECTIVES[childName];
        if (typeof directive == 'function') {
          this.emit(
            directive(child, () =>
              renderJSXNodes(
                file,
                directives,
                handlers,
                child.children,
                styles,
                {},
                false,
              ).call(this),
            ),
          );
        } else {
          if (typeof directive == 'string') {
            directives.set(childName, directive);
            if (file.module !== 'med') {
              file.import('./med.js', childName);
            }
          }
          if (isSymbol(childName)) {
            // TODO(misko): We are hard coding './med.js' which is not right.
            file.import('./med.js', childName);
            let exportedChildName = file.exports.get(childName);
            if (exportedChildName) {
              childName = exportedChildName;
            }
          }
          let props: Record<string, any> = child.properties;
          const css = child.bindings.css;
          if (css) {
            props = { ...props };
            props.class = addClass(styles.get(css)!.CLASS_NAME, props.class);
          }
          const symbolBindings: Record<string, string> = {};
          const bindings = rewriteHandlers(
            file,
            handlers,
            child.bindings,
            symbolBindings,
          );
          this.jsxBegin(childName, props, {
            ...bindings,
            ...parentSymbolBindings,
          });
          renderJSXNodes(
            file,
            directives,
            handlers,
            child.children,
            styles,
            symbolBindings,
            false,
          ).call(this);
          this.jsxEnd(childName);
        }
      }
    });
    if (needsFragment) {
      this.jsxEndFragment();
    }
    if (root) this.emit(UNINDENT, ')');
  };
}

function isSymbol(name: string): boolean {
  return name.charAt(0) == name.charAt(0).toUpperCase();
}

function addClass(
  className: string,
  existingClass: string | undefined,
): string {
  return [className, ...(existingClass ? existingClass.split(' ') : [])].join(
    ' ',
  );
}

function isEmptyTextNode(child: MitosisNode) {
  return child.properties._text?.trim() == '';
}

function isTextNode(child: MitosisNode) {
  return (
    child.properties._text !== undefined || child.bindings._text !== undefined
  );
}

/**
 * Rewrites bindings:
 * - Remove `css`
 * - Rewrites event handles
 * - Extracts symbol bindings.
 *
 * @param file
 * @param handlers
 * @param bindings
 * @param symbolBindings Options record which will receive the symbol bindings
 * @returns
 */
function rewriteHandlers(
  file: File,
  handlers: Map<string, string>,
  bindings: Record<string, string | undefined>,
  symbolBindings: Record<string, string>,
): Record<string, string> {
  const outBindings: Record<string, string> = {};
  for (let key in bindings) {
    if (Object.prototype.hasOwnProperty.call(bindings, key)) {
      let binding: any = bindings[key]!;
      let handlerBlock: string | undefined;
      if (binding != null) {
        if (key == 'css') {
          continue;
        } else if ((handlerBlock = handlers.get(binding))) {
          key = `on:${key.substring(2).toLowerCase()}`;
          binding = invoke(file.import(file.qwikModule, 'qrl'), [
            quote(file.qrlPrefix + 'high.js'),
            quote(handlerBlock),
            '[__props__, __state__]',
          ]);
        } else if (symbolBindings && key.startsWith('symbol.data.')) {
          symbolBindings[lastProperty(key)] = binding;
        }
        outBindings[key] = binding;
      }
    }
  }
  return outBindings;
}
