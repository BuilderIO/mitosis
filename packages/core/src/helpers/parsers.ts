import * as babel from '@babel/core';

const jsxPlugin = require('@babel/plugin-syntax-jsx');
const tsPreset = require('@babel/preset-typescript');
const decorators = require('@babel/plugin-syntax-decorators');

export function parseCode(code: string) {
  const ast = babel.parse(code, {
    presets: [[tsPreset, { isTSX: true, allExtensions: true }]],
    plugins: [[decorators, { legacy: true }], jsxPlugin],
  });
  const body = babel.types.isFile(ast)
    ? ast.program.body
    : babel.types.isProgram(ast)
    ? ast.body
    : [];
  return body;
}

export const isCodeBodyExpression = (body: babel.types.Statement[]) =>
  body.length == 1 &&
  (babel.types.isExpression(body[0]) ||
    babel.types.isExpressionStatement(body[0]));

/**
 * Returns `true` if the `code` is a valid expression. (vs a statement)
 */
export function isExpression(code: string): boolean {
  try {
    const body = parseCode(code);
    return isCodeBodyExpression(body);
  } catch (e) {
    return false;
  }
}

export const isCodeBodyIdentifier = (body: babel.types.Statement[]) =>
  body.length == 1 && babel.types.isIdentifier(body[0]);
