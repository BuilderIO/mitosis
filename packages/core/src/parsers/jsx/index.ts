export * from './jsx';
export { selfClosingTags } from './helpers';
export { METADATA_HOOK_NAME } from './hooks';
