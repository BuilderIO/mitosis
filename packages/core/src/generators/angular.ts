import dedent from 'dedent';
import { format } from 'prettier/standalone';
import { collectCss } from '../helpers/collect-styles';
import { fastClone } from '../helpers/fast-clone';
import { getRefs } from '../helpers/get-refs';
import { getStateObjectStringFromComponent } from '../helpers/get-state-object-string';
import { mapRefs } from '../helpers/map-refs';
import { renderPreComponent } from '../helpers/render-imports';
import { stripStateAndPropsRefs } from '../helpers/strip-state-and-props-refs';
import { selfClosingTags } from '../parsers/jsx';
import { MitosisNode } from '../types/mitosis-node';
import {
  runPostCodePlugins,
  runPostJsonPlugins,
  runPreCodePlugins,
  runPreJsonPlugins,
} from '../modules/plugins';
import isChildren from '../helpers/is-children';
import { getProps } from '../helpers/get-props';
import { kebabCase } from 'lodash';
import { stripMetaProperties } from '../helpers/strip-meta-properties';
import { removeSurroundingBlock } from '../helpers/remove-surrounding-block';
import { BaseTranspilerOptions, Transpiler } from '../types/config';
import { indent } from '../helpers/indent';

export interface ToAngularOptions extends BaseTranspilerOptions {}

const mappers: {
  [key: string]: (json: MitosisNode, options: ToAngularOptions) => string;
} = {
  Fragment: (json, options) => {
    return `<div>${json.children
      .map((item) => blockToAngular(item, options))
      .join('\n')}</div>`;
  },
};

export const blockToAngular = (
  json: MitosisNode,
  options: ToAngularOptions = {},
): string => {
  if (mappers[json.name]) {
    return mappers[json.name](json, options);
  }

  if (isChildren(json)) {
    return `<ng-content></ng-content>`;
  }

  if (json.properties._text) {
    return json.properties._text;
  }

  if (json.bindings._text) {
    return `{{${stripStateAndPropsRefs(json.bindings._text as string)}}}`;
  }

  let str = '';

  if (json.name === 'For') {
    str += `<ng-container *ngFor="let ${
      json.properties._forName
    } of ${stripStateAndPropsRefs(json.bindings.each as string)}">`;
    str += json.children
      .map((item) => blockToAngular(item, options))
      .join('\n');
    str += `</ng-container>`;
  } else if (json.name === 'Show') {
    str += `<ng-container *ngIf="${stripStateAndPropsRefs(
      json.bindings.when as string,
    )}">`;
    str += json.children
      .map((item) => blockToAngular(item, options))
      .join('\n');
    str += `</ng-container>`;
  } else {
    str += `<${json.name} `;

    // TODO: spread support for angular
    // if (json.bindings._spread) {
    //   str += `v-bind="${stripStateAndPropsRefs(
    //     json.bindings._spread as string,
    //   )}"`;
    // }

    for (const key in json.properties) {
      if (key.startsWith('$')) {
        continue;
      }
      const value = json.properties[key];
      str += ` ${key}="${value}" `;
    }
    for (const key in json.bindings) {
      if (key === '_spread') {
        continue;
      }
      if (key.startsWith('$')) {
        continue;
      }
      const value = json.bindings[key] as string;
      // TODO: proper babel transform to replace. Util for this
      const useValue = stripStateAndPropsRefs(value);

      if (key.startsWith('on')) {
        let event = key.replace('on', '').toLowerCase();
        if (
          event === 'change' &&
          json.name === 'input' /* todo: other tags */
        ) {
          event = 'input';
        }
        // TODO: proper babel transform to replace. Util for this
        const finalValue = removeSurroundingBlock(
          useValue.replace(/event\./g, '$event.'),
        );
        str += ` (${event})="${finalValue}" `;
      } else if (key === 'ref') {
        str += ` #${useValue} `;
      } else {
        str += ` [${key}]="${useValue}" `;
      }
    }
    if (selfClosingTags.has(json.name)) {
      return str + ' />';
    }
    str += '>';
    if (json.children) {
      str += json.children
        .map((item) => blockToAngular(item, options))
        .join('\n');
    }

    str += `</${json.name}>`;
  }
  return str;
};

export const componentToAngular =
  (options: ToAngularOptions = {}): Transpiler =>
  ({ component }) => {
    // Make a copy we can safely mutate, similar to babel's toolchain
    let json = fastClone(component);
    if (options.plugins) {
      json = runPreJsonPlugins(json, options.plugins);
    }

    const props = getProps(component);

    const refs = Array.from(getRefs(json));
    mapRefs(json, (refName) => `this.${refName}.nativeElement`);

    if (options.plugins) {
      json = runPostJsonPlugins(json, options.plugins);
    }
    let css = collectCss(json);
    if (options.prettier !== false) {
      css = tryFormat(css, 'css');
    }

    let template = json.children.map((item) => blockToAngular(item)).join('\n');
    if (options.prettier !== false) {
      template = tryFormat(template, 'html');
    }

    stripMetaProperties(json);

    const dataString = getStateObjectStringFromComponent(json, {
      format: 'class',
      valueMapper: (code) =>
        stripStateAndPropsRefs(code, { replaceWith: 'this.' }),
    });

    let str = dedent`
    import { Component ${refs.length ? ', ViewChild, ElementRef' : ''}${
      props.size ? ', Input' : ''
    } } from '@angular/core';
    ${renderPreComponent(json)}

    @Component({
      selector: '${kebabCase(json.name || 'my-component')}',
      template: \`
        ${indent(template, 8).replace(/`/g, '\\`').replace(/\$\{/g, '\\${')}
      \`,
      ${
        css.length
          ? `styles: [
        \`${indent(css, 8)}\`
      ],`
          : ''
      }
    })
    export default class ${component.name} {
      ${Array.from(props)
        .map((item) => `@Input() ${item}: any`)
        .join('\n')}

      ${refs
        .map((refName) => `@ViewChild('${refName}') ${refName}: ElementRef`)
        .join('\n')}

      ${
        !component.hooks.onMount
          ? ''
          : `ngOnInit() {
              ${stripStateAndPropsRefs(component.hooks.onMount.code, {
                replaceWith: 'this.',
              })}
            }`
      }

      ${
        !component.hooks.onUpdate?.length
          ? ''
          : `ngAfterContentChecked() {
              ${component.hooks.onUpdate.map((hook) =>
                stripStateAndPropsRefs(hook.code, {
                  replaceWith: 'this.',
                }),
              )}
            }`
      }

      ${
        !component.hooks.onUnMount
          ? ''
          : `ngOnDestroy() {
              ${stripStateAndPropsRefs(component.hooks.onUnMount.code, {
                replaceWith: 'this.',
              })}
            }`
      }

      ${dataString}
    }
  `;

    if (options.plugins) {
      str = runPreCodePlugins(str, options.plugins);
    }
    if (options.prettier !== false) {
      str = tryFormat(str, 'typescript');
    }
    if (options.plugins) {
      str = runPostCodePlugins(str, options.plugins);
    }
    return str;
  };

const tryFormat = (str: string, parser: string) => {
  try {
    return format(str, {
      parser,
      plugins: [
        // To support running in browsers
        require('prettier/parser-typescript'),
        require('prettier/parser-postcss'),
        require('prettier/parser-html'),
        require('prettier/parser-babel'),
      ],
      htmlWhitespaceSensitivity: 'ignore',
    });
  } catch (err) {
    console.warn('Could not prettify', { string: str }, err);
  }
  return str;
};
