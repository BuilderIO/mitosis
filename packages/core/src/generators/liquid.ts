import { format } from 'prettier/standalone';
import { collectCss } from '../helpers/collect-styles';
import { fastClone } from '../helpers/fast-clone';
import { stripStateAndPropsRefs } from '../helpers/strip-state-and-props-refs';
import { selfClosingTags } from '../parsers/jsx';
import { JSXLiteComponent } from '../types/jsx-lite-component';
import { JSXLiteNode } from '../types/jsx-lite-node';
import {
  Plugin,
  runPostCodePlugins,
  runPostJsonPlugins,
  runPreCodePlugins,
  runPreJsonPlugins,
} from '../modules/plugins';

/**
 * Test if the binding expression would be likely to generate
 * valid or invalid liquid. If we generate invalid liquid tags
 * Shopify will reject our PUT to update the template
 */
export const isValidLiquidBinding = (str = '') => {
  const strictMatches = Boolean(
    // Test for our `context.shopify.liquid.*(expression), which
    // we regex out later to transform back into valid liquid expressions
    str.match(/(context|ctx)\s*(\.shopify\s*)?\.liquid\s*\./),
  );

  return (
    strictMatches ||
    // Test is the expression is simple and would map to Shopify bindings	    // Test for our `context.shopify.liquid.*(expression), which
    // e.g. `state.product.price` -> `{{product.price}}	    // we regex out later to transform back into valid liquid expressions
    Boolean(str.match(/^[a-z0-9_\.\s]+$/i))
  );
};

type ToLiquidOptions = {
  prettier?: boolean;
  plugins?: Plugin[];
};

const mappers: {
  [key: string]: (json: JSXLiteNode, options: ToLiquidOptions) => string;
} = {
  Fragment: (json, options) => {
    return `<div>${json.children
      .map((item) => blockToLiquid(item, options))
      .join('\n')}</div>`;
  },
};

// TODO: spread support
const blockToLiquid = (
  json: JSXLiteNode,
  options: ToLiquidOptions = {},
): string => {
  if (mappers[json.name]) {
    return mappers[json.name](json, options);
  }

  // TODO: Add support for `{props.children}` bindings

  if (json.properties._text) {
    return json.properties._text;
  }

  if (json.bindings._text) {
    if (!isValidLiquidBinding(json.bindings._text as string)) {
      return '';
    }
    return `{{${stripStateAndPropsRefs(json.bindings._text as string)}}}`;
  }

  let str = '';

  if (json.name === 'For') {
    if (
      !(
        isValidLiquidBinding(json.bindings.each as string) &&
        isValidLiquidBinding(json.bindings._forName as string)
      )
    ) {
      return str;
    }
    str += `{% for ${json.bindings._forName} in ${stripStateAndPropsRefs(
      json.bindings.each as string,
    )} %}`;
    if (json.children) {
      str += json.children
        .map((item) => blockToLiquid(item, options))
        .join('\n');
    }

    str += '{% endfor %}';
  } else if (json.name === 'Show') {
    if (!isValidLiquidBinding(json.bindings.when as string)) {
      return str;
    }
    str += `{% if ${stripStateAndPropsRefs(json.bindings.when as string)} %}`;
    if (json.children) {
      str += json.children
        .map((item) => blockToLiquid(item, options))
        .join('\n');
    }

    str += '{% endif %}';
  } else {
    str += `<${json.name} `;

    if (
      json.bindings._spread === '_spread' &&
      isValidLiquidBinding(json.bindings._spread)
    ) {
      str += `
          {% for _attr in ${json.bindings._spread} %}
            {{ _attr[0] }}="{{ _attr[1] }}"
          {% endfor %}
        `;
    }

    for (const key in json.properties) {
      const value = json.properties[key];
      str += ` ${key}="${value}" `;
    }

    for (const key in json.bindings) {
      if (key === '_spread' || key === 'ref' || key === 'css') {
        continue;
      }
      const value = json.bindings[key] as string;
      // TODO: proper babel transform to replace. Util for this
      const useValue = stripStateAndPropsRefs(value);

      if (key.startsWith('on')) {
        // Do nothing
      } else if (isValidLiquidBinding(useValue)) {
        str += ` ${key}="{{${useValue}}}" `;
      }
    }
    if (selfClosingTags.has(json.name)) {
      return str + ' />';
    }
    str += '>';
    if (json.children) {
      str += json.children
        .map((item) => blockToLiquid(item, options))
        .join('\n');
    }

    str += `</${json.name}>`;
  }
  return str;
};

// TODO: add JS support similar to componentToHtml()
export const componentToLiquid = (
  componentJson: JSXLiteComponent,
  options: ToLiquidOptions = {},
) => {
  let json = fastClone(componentJson);
  if (options.plugins) {
    json = runPreJsonPlugins(json, options.plugins);
  }
  const css = collectCss(json);
  if (options.plugins) {
    json = runPostJsonPlugins(json, options.plugins);
  }
  let str = json.children.map((item) => blockToLiquid(item)).join('\n');

  if (css.trim().length) {
    str += `<style>${css}</style>`;
  }

  if (options.plugins) {
    str = runPreCodePlugins(str, options.plugins);
  }
  if (options.prettier !== false) {
    try {
      str = format(str, {
        parser: 'html',
        htmlWhitespaceSensitivity: 'ignore',
        plugins: [
          // To support running in browsers
          require('prettier/parser-html'),
          require('prettier/parser-postcss'),
          require('prettier/parser-babel'),
        ],
      });
    } catch (err) {
      console.warn('Could not prettify', { string: str }, err);
    }
  }
  if (options.plugins) {
    str = runPostCodePlugins(str, options.plugins);
  }
  return str;
};
