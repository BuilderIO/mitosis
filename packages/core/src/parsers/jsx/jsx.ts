import * as babel from '@babel/core';
import generate from '@babel/generator';
import { pipe } from 'fp-ts/lib/function';
import { createMitosisComponent } from '../../helpers/create-mitosis-component';
import { tryParseJson } from '../../helpers/json';
import { stripNewlinesInStrings } from '../../helpers/replace-new-lines-in-strings';
import { MitosisComponent } from '../../types/mitosis-component';
import { jsonToAst } from './ast';
import { collectTypes, isTypeOrInterface } from './component-types';
import { extractContextComponents } from './context';
import { jsxElementToJson } from './element-parser';
import { generateExports } from './exports';
import { componentFunctionToJson } from './function-parser';
import { isImportOrDefaultExport } from './helpers';
import { collectModuleScopeHooks } from './hooks';
import { handleImportDeclaration } from './imports';
import { undoPropsDestructure } from './props';
import { mapStateIdentifiers } from './state';
import { Context, ParseMitosisOptions } from './types';

import tsPlugin from '@babel/plugin-syntax-typescript';
import tsPreset from '@babel/preset-typescript';
import { HOOKS } from '../../constants/hooks';
import { getMagicString, getTargetId, getUseTargetStatements } from './hooks/use-target';

const { types } = babel;

const typescriptBabelPreset = [tsPreset, { isTSX: true, allExtensions: true }];

const beforeParse = (path: babel.NodePath<babel.types.Program>) => {
  path.traverse({
    FunctionDeclaration(path) {
      undoPropsDestructure(path);
    },
  });
};

/**
 * This function takes the raw string from a Mitosis component, and converts it into a JSON that can be processed by
 * each generator function.
 *
 * @param jsx string representation of the Mitosis component
 * @returns A JSON representation of the Mitosis component
 */
export function parseJsx(
  jsx: string,
  _options: Partial<ParseMitosisOptions> = {},
): MitosisComponent {
  let subComponentFunctions: string[] = [];

  const options: ParseMitosisOptions = {
    typescript: false,
    ..._options,
  };

  const jsxToUse = options.typescript
    ? jsx
    : // strip typescript types by running through babel's TS preset.
      (babel.transform(jsx, {
        configFile: false,
        babelrc: false,
        presets: [typescriptBabelPreset],
      })?.code as string);

  const output = babel.transform(jsxToUse, {
    configFile: false,
    babelrc: false,
    comments: false,
    plugins: [
      [tsPlugin, { isTSX: true }],
      (): babel.PluginObj<Context> => ({
        visitor: {
          JSXExpressionContainer(path, context) {
            if (types.isJSXEmptyExpression(path.node.expression)) {
              path.remove();
            }
          },
          Program(path, context) {
            if (context.builder) {
              return;
            }

            beforeParse(path);

            context.builder = {
              component: createMitosisComponent(),
            };

            const keepStatements = path.node.body.filter(
              (statement) => isImportOrDefaultExport(statement) || isTypeOrInterface(statement),
            );

            context.builder.component.exports = generateExports(path);

            subComponentFunctions = path.node.body
              .filter(
                (node) =>
                  !types.isExportDefaultDeclaration(node) && types.isFunctionDeclaration(node),
              )
              .map((node) => `export default ${generate(node).code!}`);

            const preComponentCode = pipe(
              path.node.body.filter((statement) => !isImportOrDefaultExport(statement)),
              collectModuleScopeHooks(context.builder.component, options),
              types.program,
              generate,
              (generatorResult) => generatorResult.code,
            );

            // TODO: support multiple? e.g. for others to add imports?
            context.builder.component.hooks.preComponent = { code: preComponentCode };

            path.replaceWith(types.program(keepStatements));
          },
          FunctionDeclaration(path, context) {
            const { node } = path;
            if (types.isIdentifier(node.id)) {
              const name = node.id.name;
              if (name[0].toUpperCase() === name[0]) {
                path.traverse({
                  /**
                   * Plugin to find all `useTarget()` assignment calls inside of the component function body
                   * and replace them with a magic string.
                   */
                  CallExpression(path) {
                    if (!types.isVariableDeclarator(path.parent)) return;
                    if (!types.isCallExpression(path.node)) return;
                    if (!types.isIdentifier(path.node.callee)) return;
                    if (path.node.callee.name !== HOOKS.TARGET) return;

                    const targetBlock = getUseTargetStatements(path.node);

                    if (!targetBlock) return;

                    const blockId = getTargetId(context.builder.component);

                    // replace the useTarget() call with a magic string
                    path.replaceWith(types.stringLiteral(getMagicString(blockId)));

                    // store the target block in the component
                    context.builder.component.targetBlocks = {
                      ...context.builder.component.targetBlocks,
                      [blockId]: targetBlock,
                    };
                  },
                });
                path.replaceWith(jsonToAst(componentFunctionToJson(node, context)));
              }
            }
          },
          ImportDeclaration(path, context) {
            handleImportDeclaration({ options, path, context });
          },
          ExportDefaultDeclaration(path) {
            path.replaceWith(path.node.declaration);
          },
          JSXElement(path) {
            const { node } = path;
            path.replaceWith(jsonToAst(jsxElementToJson(node)));
          },
          ExportNamedDeclaration(path, context) {
            const { node } = path;
            if (
              babel.types.isTSInterfaceDeclaration(node.declaration) ||
              babel.types.isTSTypeAliasDeclaration(node.declaration)
            ) {
              collectTypes(path, context);
            }
          },
          TSTypeAliasDeclaration(path, context) {
            collectTypes(path, context);
          },
          TSInterfaceDeclaration(path, context) {
            collectTypes(path, context);
          },
        },
      }),
    ],
  });

  const stringifiedMitosisComponent = stripNewlinesInStrings(
    output!
      .code!.trim()
      // Occasional issues where comments get kicked to the top. Full fix should strip these sooner
      .replace(/^\/\*[\s\S]*?\*\/\s*/, '')
      // Weird bug with adding a newline in a normal at end of a normal string that can't have one
      // If not one-off find full solve and cause
      .replace(/\n"/g, '"')
      .replace(/^\({/, '{')
      .replace(/}\);$/, '}'),
  );

  const mitosisComponent = tryParseJson(stringifiedMitosisComponent);

  mapStateIdentifiers(mitosisComponent);
  extractContextComponents(mitosisComponent);

  mitosisComponent.subComponents = subComponentFunctions.map((item) => parseJsx(item, options));

  return mitosisComponent;
}
