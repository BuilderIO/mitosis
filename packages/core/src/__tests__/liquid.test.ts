import { componentToLiquid } from '../generators/liquid';
import { parseJsx } from '../parsers/jsx';
const basic = require('./data/basic.raw');
const submitButtonBlock = require('./data/blocks/submit-button.raw');
const inputBlock = require('./data/blocks/input.raw');
const selectBlock = require('./data/blocks/select.raw');
const formBlock = require('./data/blocks/form.raw');
const button = require('./data/blocks/button.raw');
const textarea = require('./data/blocks/textarea.raw');
const img = require('./data/blocks/img.raw');
const video = require('./data/blocks/video.raw');
const section = require('./data/blocks/section.raw');
const text = require('./data/blocks/text.raw');

describe('Liquid', () => {
  test('Basic', () => {
    const json = parseJsx(basic);
    const output = componentToLiquid(json);
    expect(output).toMatchSnapshot();
  });

  test('Input block', () => {
    const json = parseJsx(inputBlock);
    const output = componentToLiquid(json);
    expect(output).toMatchSnapshot();
  });

  test('Submit button block', () => {
    const json = parseJsx(submitButtonBlock);
    const output = componentToLiquid(json);
    expect(output).toMatchSnapshot();
  });

  test('Select block', () => {
    const json = parseJsx(selectBlock);
    const output = componentToLiquid(json);
    expect(output).toMatchSnapshot();
  });

  test('Form block', () => {
    const json = parseJsx(formBlock);
    const output = componentToLiquid(json);
    expect(output).toMatchSnapshot();
  });

  test('Button', () => {
    const json = parseJsx(button);
    const output = componentToLiquid(json);
    expect(output).toMatchSnapshot();
  });

  test('Textarea', () => {
    const json = parseJsx(textarea);
    const output = componentToLiquid(json);
    expect(output).toMatchSnapshot();
  });

  test('Img', () => {
    const json = parseJsx(img);
    const output = componentToLiquid(json);
    expect(output).toMatchSnapshot();
  });

  test('Video', () => {
    const json = parseJsx(video);
    const output = componentToLiquid(json);
    expect(output).toMatchSnapshot();
  });

  test('Section', () => {
    const json = parseJsx(section);
    const output = componentToLiquid(json);
    expect(output).toMatchSnapshot();
  });

  test('Text', () => {
    const json = parseJsx(text);
    const output = componentToLiquid(json);
    expect(output).toMatchSnapshot();
  });
});
