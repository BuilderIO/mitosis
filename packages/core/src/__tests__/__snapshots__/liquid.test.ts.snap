// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Liquid Basic 1`] = `
"<div>
  <input value=\\"{{name}}\\" />

  Hello! I can run in React, Vue, Solid, or Liquid!
</div>
"
`;

exports[`Liquid Button 1`] = `
"<div>
  {% if link %}

  <a href=\\"{{link}}\\">{{text}}</a>

  {% endif %}
</div>
"
`;

exports[`Liquid Columns 1`] = `
"<div class=\\"builder-columns div\\">
  {% for column in columns %}
  <div class=\\"builder-column div-2\\">{{column.content}}</div>
  {% endfor %}
</div>
<style>
  .div {
    display: flex;
    flex-direction: column;
    align-items: stretch;
    line-height: normal;
  }
  @media (max-width: 999px) {
    .div {
      flex-direction: row;
    }
  }
  @media (max-width: 639px) {
    .div {
      flex-direction: row-reverse;
    }
  }
  .div-2 {
    flex-grow: 1;
  }
</style>
"
`;

exports[`Liquid CustomCode 1`] = `
"<div innerHTML=\\"{{code}}\\"></div>
"
`;

exports[`Liquid Embed 1`] = `
"<div innerHTML=\\"{{code}}\\"></div>
"
`;

exports[`Liquid Form block 1`] = `
"<form validate=\\"{{validate}}\\" method=\\"{{method}}\\" name=\\"{{name}}\\"></form>
<style>
  .pre {
    padding: 10px;
    color: red;
    text-align: center;
  }
</style>
"
`;

exports[`Liquid Image 1`] = `
"<div>
  <picture>
    <source srcset=\\"{{srcset}}\\" />
  </picture>

  {{children}}
</div>
<style>
  .img {
    opacity: 1;
    transition: opacity 0.2s ease-in-out;
    object-fit: cover;
    object-position: center;
  }
</style>
"
`;

exports[`Liquid Img 1`] = `
"<img alt=\\"{{altText}}\\" src=\\"{{imgSrc}}\\" />
"
`;

exports[`Liquid Input block 1`] = `
"<input
  placeholder=\\"{{placeholder}}\\"
  type=\\"{{type}}\\"
  name=\\"{{name}}\\"
  value=\\"{{value}}\\"
  defaultValue=\\"{{defaultValue}}\\"
  required=\\"{{required}}\\"
/>
"
`;

exports[`Liquid RawText 1`] = `
"<span></span>
"
`;

exports[`Liquid Section 1`] = `
"<section>{{children}}</section>
"
`;

exports[`Liquid Select block 1`] = `
"<select value=\\"{{value}}\\" defaultValue=\\"{{defaultValue}}\\" name=\\"{{name}}\\">
  {% for option in options %}
  <option value=\\"{{option.value}}\\"></option>
  {% endfor %}
</select>
"
`;

exports[`Liquid Stamped.io 1`] = `
"<div>
  <button>Write a review</button>

  {% if showReviewPrompt %}

  <input placeholder=\\"Email\\" />

  <input placeholder=\\"Title\\" class=\\"input\\" />

  <textarea placeholder=\\"How was your experience?\\" class=\\"textarea\\"></textarea>

  <button class=\\"button\\">Submit</button>

  {% endif %} {% for review in reviews %}
  <div class=\\"review\\" key=\\"{{review.id}}\\">
    <img class=\\"img\\" src=\\"{{review.avatar}}\\" />

    <div>
      <div>{{review.author}}</div>

      <div>{{review.reviewMessage}}</div>
    </div>
  </div>
  {% endfor %}
</div>
<style>
  .input {
    display: block;
  }
  .textarea {
    display: block;
  }
  .button {
    display: block;
  }
  .review {
    margin: 10px;
    padding: 10px;
    background: white;
    display: flex;
    border-radius: 5px;
    box-shadow: 0 2px 5px rgba(0, 0, 0, 0.1);
  }
  .img {
    height: 30px;
    width: 30px;
    margin-right: 10px;
  }
</style>
"
`;

exports[`Liquid Submit button block 1`] = `
"<button type=\\"submit\\">{{text}}</button>
"
`;

exports[`Liquid Text 1`] = `
"<div></div>
"
`;

exports[`Liquid Textarea 1`] = `
"<textarea
  placeholder=\\"{{placeholder}}\\"
  name=\\"{{name}}\\"
  value=\\"{{value}}\\"
  defaultValue=\\"{{defaultValue}}\\"
></textarea>
"
`;

exports[`Liquid Video 1`] = `
"<video
  preload=\\"none\\"
  poster=\\"{{posterImage}}\\"
  autoplay=\\"{{autoPlay}}\\"
  muted=\\"{{muted}}\\"
  controls=\\"{{controls}}\\"
  loop=\\"{{loop}}\\"
></video>
"
`;
